dashboard/app/models/concerns/partial_registration.rb

```diff
@@ -8,16 +8,64 @@ module PartialRegistration
   extend ActiveSupport::Concern
 
   USER_ATTRIBUTES_SESSION_KEY = 'devise.user_attributes'
-  OAUTH_PARAMS_TO_STRIP = %w(oauth_token oauth_refresh_token)
+
+  # A common error during registration, especially when OAuth is involved, is a
+  # Cookie Overflow.  This module stores attributes for a partially-created user
+  # in the _learn_session cookie, which cannot exceed 4KB.  In fact, it we seem
```

Comment:
Good point, you've convinced me that even if we consider changing our session storage implementation, we should still do something now on the codepath addressed in this PR to help fix this problem in the short term.

Thinking more, because cookies are stored on the client, it's not even clear to me how we could really "migrate" away from them without maintaining two session storage implementations for a long period of time, which (unless rails or some gem supports this) doesn't sound practical.

Reply:
@wjordan and I discussed this last recently. Our usage is likely too high to use the database (active record store), but memcached might be a good option for us.

Reply:
> Thinking more, because cookies are stored on the client, it's not even clear to me how we could really "migrate" away from them without maintaining two session storage implementations for a long period of time, which (unless rails or some gem supports this) doesn't sound practical.

I did a little digging on this and didn't find anything battle-hardened.  It looks like most teams pick a session store early and stick with it, and many of those that go through a change accept the one-time cost of dropping all user sessions.

For a lot of applications, dropping sessions just means that users need to sign in again.  In our case, it's not just that we'd rather not inconvenience our users by signing them out.  We're also doing quite a bit with our session for signed-out users, which would be lost in this scenario - and we know that signed-out users are a large part of our audience.

I did find scattered reports of teams putting code in their [ApplicationController](https://stackoverflow.com/a/40118169/5000129) or [SessionController](https://stackoverflow.com/a/52413056/5000129) for a few weeks to migrate active sessions.  I suspect 4 weeks would be enough to significantly mitigate the impact of such a change for our application (especially if done over the summer) but we would want to be very careful, since it seems easy to introduce security issues or mangle sessions if we're writing our own code to handle them.

Reply:
[This article from Justin Weiss](https://www.justinweiss.com/articles/how-rails-sessions-work/) goes into more detail on the downsides of cache-based session stores.

> **Storing sessions in the cache**
>
> You might already be using something like Memcache to cache your partials or data. If so, the cache store is the second-easiest place to store session data, since it’s already set up.
> 
> You don’t have to worry about your session store growing out of control, because older sessions will automatically get kicked out of the cache if it gets too big. And it’s fast, because your cache will most likely be kept in memory.
> 
> But it’s not perfect:
> 
> - If you actually care about keeping old sessions around, you probably don’t want them to get kicked out of the cache.
> 
> - Your sessions and your cached data will be fighting for space. If you don’t have enough memory, you could be facing a ton of cache misses and early expired sessions.
> 
> - If you ever need to reset your cache (say you upgraded Rails and your old cached data is no longer accurate), there’s no way to do that without expiring everyone’s sessions.
> 
> Still, this is how we store session data at Avvo, and it’s worked well for us so far.

A better version of this design might use a dedicated cache cluster for sessions, separate from the application cache.  You wouldn't fight with the application cache for space or drop sessions when you flush the application cache, and you could specifically configure your session cache for [LRU replacement](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)) to ensure the sessions you _do_ drop are the least active ones.  Still, dropped sessions at some scale seems like an inevitable consequence of moving session contents to our servers.

> By default Rails is just using encrypted cookies on the client. This eliminates the single point of failure completely. This approach is the best in most cases.
> [Kirill Shevchenko, 2018](https://medium.com/@kirill_shevch/configuration-cache-and-rails-session-store-with-redis-b3ce6f64d1fc)

After reading about session storage options some more and seeing several "best in most cases" opinions like the one above, I'm convinced that **we are misusing the session** specifically in the cases of PartialRegistration and callouts_seen, and offloading these cases to a more appropriate storage solution would resolve the majority of our CookieOverflow issues.  Working out an improved design for these two components is going to take some work, but my instinct is that it's still less expensive than the risk, effort, and maintenance cost of migrating to a different session store.

Reply:
> After reading about session storage options some more and seeing several "best in most cases" opinions like the one above, I'm convinced that **we are misusing the session** specifically in the cases of PartialRegistration and callouts_seen, and offloading these cases to a more appropriate storage solution would resolve the majority of our CookieOverflow issues. Working out an improved design for these two components is going to take some work, but my instinct is that it's still less expensive than the risk, effort, and maintenance cost of migrating to a different session store.

+ 💯 !